File: nidra/collector/dreamer.mjs
// The Dreamer — Nidra's "Sleep" pass. Where deriveOpinions (heuristic) counts
// tokens, the dreamer asks an LLM to CONNECT signals across the day into
// higher-level intent and a model of the person (e.g. flights + car rental +
// hotel reading = "planning a trip"). Default engine: on-device Gemma 4 via
// Ollama (privacy-first). `callModel` is injectable so tests stay deterministic.

const OLLAMA_HOST = process.env.OLLAMA_HOST || "http://localhost:11434";
const DREAM_MODEL = process.env.NIDRA_DREAM_MODEL || "gemma4:latest";

const SYSTEM = `You are Nidra's "dreamer". While the user sleeps you consolidate their recent online activity.
Your job is to CONNECT THE DOTS ACROSS signals into higher-level understanding — NOT to repeat the activity back.
Look hard for latent intent that spans MULTIPLE signals (e.g. a flight search + a car-rental search + reading about hotels = planning a trip; repeated reading on one topic + a related search = an active project).
Infer interests, reading taste, active projects, and the user's likely next needs. Be specific and confident when several signals corroborate; hedge when the evidence is thin. Never invent activity that isn't in the digest.
Respond with ONLY a JSON object of this exact shape:
{
  "connectedInsights": [{"insight": "one clear sentence", "fromSignals": ["signal", "signal"], "confidence": 0.0, "reasoning": "why these connect"}],
  "persona": "2-3 sentence portrait of who this person is right now",
  "beliefs": [{"statement": "a durable belief about the user", "confidence": 0.0}],
  "nextNeeds": ["a proactive thing Nidra could do next"]
}`;

/** Build the compact, already-redacted text the dreamer reads. */
export function digest(events = []) {
  const by = (t) => events.filter((e) => e.type === t);
  const lines = [];
  const searches = by("search").map((e) => e.data?.query).filter(Boolean);
  if (searches.length) lines.push("SEARCHES:\n" + searches.map((q) => `- ${q}`).join("\n"));

  const reading = by("reading").map((e) => {
    const tags = (e.data?.tags || []).slice(0, 5).join(", ");
    return `- "${e.data?.title || e.title || e.domain}"${e.data?.author ? " by " + e.data.author : ""}${tags ? " (" + tags + ")" : ""} [${e.source}]`;
  });
  if (reading.length) lines.push("READING:\n" + reading.join("\n"));

  const emails = by("email").map((e) => `- ${e.data?.action || "mail"}: ${e.data?.subject || "(no subject)"}`);
  if (emails.length) lines.push("EMAIL:\n" + emails.join("\n"));

  const cal = by("calendar").map((e) => `- ${e.data?.action || "view"}: ${e.data?.eventTitle || "(event)"}`);
  if (cal.length) lines.push("CALENDAR:\n" + cal.join("\n"));

  const domains = {};
  for (const e of events) if (e.domain) domains[e.domain] = (domains[e.domain] || 0) + 1;
  const top = Object.entries(domains).sort((a, b) => b[1] - a[1]).slice(0, 8).map(([d]) => d);
  if (top.length) lines.push("TOP SITES: " + top.join(", "));

  return lines.join("\n\n") || "(no activity captured)";
}

async function ollamaChat(prompt, { host, model }) {
  const res = await fetch(host + "/api/chat", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      model,
      stream: false,
      format: "json",
      options: { temperature: 0.3 },
      messages: [
        { role: "system", content: SYSTEM },
        { role: "user", content: prompt },
      ],
    }),
  });
  if (!res.ok) throw new Error(`ollama ${res.status}`);
  const data = await res.json();
  return data?.message?.content || "";
}

/** Robustly pull a JSON object out of a model response. */
export function extractJson(text) {
  if (!text) return {};
  let t = String(text).trim().replace(/^```(?:json)?/i, "").replace(/```$/, "").trim();
  try {
    return JSON.parse(t);
  } catch {
    const s = t.indexOf("{");
    const e = t.lastIndexOf("}");
    if (s >= 0 && e > s) {
      try {
        return JSON.parse(t.slice(s, e + 1));
      } catch {
        /* fall through */
      }
    }
    return {};
  }
}

/**
 * Dream over a list of activity events.
 * @param events captured ActivityEvents
 * @param opts.callModel async (prompt) => string — override the engine (tests)
 */
export async function dream(events = [], { callModel, host = OLLAMA_HOST, model = DREAM_MODEL } = {}) {
  const prompt = digest(events);
  const call = callModel || ((p) => ollamaChat(p, { host, model }));
  const parsed = extractJson(await call(prompt));
  return {
    engine: callModel ? "mock" : `ollama:${model}`,
    generatedFrom: events.length,
    connectedInsights: parsed.connectedInsights || parsed.insights || [],
    persona: parsed.persona || null,
    beliefs: parsed.beliefs || [],
    nextNeeds: parsed.nextNeeds || parsed.next_needs || [],
  };
}

/** Is an Ollama-style engine reachable? (used to gate live tests / UI) */
export async function dreamerAvailable({ host = OLLAMA_HOST } = {}) {
  try {
    const res = await fetch(host + "/api/tags", { signal: AbortSignal.timeout(2000) });
    return res.ok;
  } catch {
    return false;
  }
}

File: backend/src/pragya_assistant/connectors/browser_activity/dreamer.py
"""The dreamer — Nidra's LLM "Sleep" pass over ingested browser activity.

Where the store just records events, the dreamer asks an LLM to CONNECT signals
across them into higher-level intent + a persona + next-needs (e.g. flight search
+ car-rental search + hotel reading = "planning a trip"). Default engine:
on-device Gemma 4 via Ollama. The completion callable is injected so tests stay
deterministic — no network.
"""

from __future__ import annotations

import json
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from pragya_assistant.connectors.browser_activity.store import BrowserActivityEventStore
    from pragya_assistant.memory.models import BrowserActivityEvent

# A completion callable: (user digest) -> raw model text (expected to be JSON).
DreamFn = Callable[[str], Awaitable[str]]

SYSTEM = (
    'You are Nidra\'s "dreamer". While the user sleeps you consolidate their recent online '
    "activity. Your job is to CONNECT THE DOTS ACROSS signals into higher-level understanding "
    "— NOT to repeat the activity back. Look hard for latent intent that spans MULTIPLE signals "
    "(e.g. a flight search + a car-rental search + reading about hotels = planning a trip; "
    "repeated reading on one topic + a related search = an active project). Infer interests, "
    "reading taste, active projects, and the user's likely next needs. Be specific and confident "
    "when several signals corroborate; hedge when the evidence is thin. Never invent activity that "
    "isn't in the digest. Respond with ONLY a JSON object of this exact shape: "
    '{"connectedInsights": [{"insight": "one clear sentence", "fromSignals": ["signal"], '
    '"confidence": 0.0, "reasoning": "why these connect"}], "persona": "2-3 sentence portrait of '
    'who this person is right now", "beliefs": [{"statement": "a durable belief", "confidence": '
    '0.0}], "nextNeeds": ["a proactive thing Nidra could do next"]}'
)


def build_digest(events: list[BrowserActivityEvent]) -> str:
    """The compact, already-redacted text the dreamer reads."""

    def of(event_type: str) -> list[BrowserActivityEvent]:
        return [e for e in events if e.event_type == event_type]

    lines: list[str] = []

    searches = [
        str((e.data or {}).get("query")) for e in of("search") if (e.data or {}).get("query")
    ]
    if searches:
        lines.append("SEARCHES:\n" + "\n".join(f"- {q}" for q in searches))

    reading: list[str] = []
    for e in of("reading"):
        d = e.data or {}
        tags = ", ".join(str(t) for t in (d.get("tags") or [])[:5])
        title = d.get("title") or e.title or e.domain or "(page)"
        author = f" by {d['author']}" if d.get("author") else ""
        suffix = f" ({tags})" if tags else ""
        reading.append(f'- "{title}"{author}{suffix} [{e.source}]')
    if reading:
        lines.append("READING:\n" + "\n".join(reading))

    emails = [
        f"- {(e.data or {}).get('action', 'mail')}: {(e.data or {}).get('subject', '(no subject)')}"
        for e in of("email")
    ]
    if emails:
        lines.append("EMAIL:\n" + "\n".join(emails))

    calendar = [
        f"- {(e.data or {}).get('action', 'view')}: {(e.data or {}).get('eventTitle', '(event)')}"
        for e in of("calendar")
    ]
    if calendar:
        lines.append("CALENDAR:\n" + "\n".join(calendar))

    domains: dict[str, int] = {}
    for e in events:
        if e.domain:
            domains[e.domain] = domains.get(e.domain, 0) + 1
    top = sorted(domains, key=lambda d: domains[d], reverse=True)[:8]
    if top:
        lines.append("TOP SITES: " + ", ".join(top))

    return "\n\n".join(lines) or "(no activity captured)"


def extract_json(text: str) -> dict[str, Any]:
    """Robustly pull a JSON object out of a model response."""
    if not text:
        return {}
    stripped = text.strip().removeprefix("```json").removeprefix("```").removesuffix("```").strip()
    try:
        parsed = json.loads(stripped)
    except json.JSONDecodeError:
        start, end = stripped.find("{"), stripped.rfind("}")
        if start < 0 or end <= start:
            return {}
        try:
            parsed = json.loads(stripped[start : end + 1])
        except json.JSONDecodeError:
            return {}
    return parsed if isinstance(parsed, dict) else {}


@dataclass(frozen=True)
class DreamResult:
    generated_from: int
    persona: str | None
    connected_insights: list[dict[str, Any]]
    beliefs: list[dict[str, Any]]
    next_needs: list[str]
    engine: str


def _as_list(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


class DreamerService:
    """Reads recent activity, asks the LLM to connect the dots, returns a dream."""

    def __init__(
        self,
        store: BrowserActivityEventStore,
        complete: DreamFn,
        *,
        connector_key: str = "browser_activity",
        engine_label: str = "mock",
    ) -> None:
        self._store = store
        self._complete = complete
        self._key = connector_key
        self._engine_label = engine_label

    async def dream(self, *, limit: int = 200) -> DreamResult:
        events = await self._store.recent(self._key, limit=limit)
        parsed = extract_json(await self._complete(build_digest(events)))
        persona = parsed.get("persona")
        return DreamResult(
            generated_from=len(events),
            persona=persona if isinstance(persona, str) else None,
            connected_insights=_as_list(parsed.get("connectedInsights") or parsed.get("insights")),
            beliefs=_as_list(parsed.get("beliefs")),
            next_needs=[str(n) for n in _as_list(parsed.get("nextNeeds"))],
            engine=self._engine_label,
        )


def ollama_dream_fn(base_url: str, model: str, *, timeout: float = 120.0) -> DreamFn:
    """The default engine: on-device Gemma 4 via Ollama with forced JSON output."""

    async def _call(digest: str) -> str:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.post(
                f"{base_url.rstrip('/')}/api/chat",
                json={
                    "model": model,
                    "stream": False,
                    "format": "json",
                    "options": {"temperature": 0.3},
                    "messages": [
                        {"role": "system", "content": SYSTEM},
                        {"role": "user", "content": digest},
                    ],
                },
            )
            resp.raise_for_status()
            data: dict[str, Any] = resp.json()
            message = data.get("message") or {}
            return str(message.get("content") or "")

    return _call
